src/stdlib_sorting_sort_index.fypp

```diff
@@ -0,0 +1,1684 @@
+! copying the lesser (or greater) one into `array`.
+
+            if ( mid <= array_len - mid ) then ! The left run is shorter.
+                buf(0:mid-1) = array(0:mid-1)
+                ibuf(0:mid-1) = index(0:mid-1)
+                i = 0
+                j = mid
+                merge_lower: do k = 0, array_len-1
+                    if ( buf(i) <= array(j) ) then
+                        array(k) = buf(i)
+                        index(k) = ibuf(i)
+                        i = i + 1
+                        if ( i >= mid ) exit merge_lower
+                    else
+                        array(k) = array(j)
+                        index(k) = index(j)
+                        j = j + 1
+                        if ( j >= array_len ) then
+                            array(k+1:) = buf(i:mid-1)
+                            index(k+1:) = ibuf(i:mid-1)
+                            exit merge_lower
+                        end if
+                    end if
+                end do merge_lower
+            else ! The right run is shorter
+                buf(0:array_len-mid-1) = array(mid:array_len-1)
+                ibuf(0:array_len-mid-1) = index(mid:array_len-1)
+                i = mid - 1
+                j = array_len - mid -1
+                merge_upper: do k = array_len-1, 0, -1
+                    if ( buf(j) >= array(i) ) then
+                        array(k) = buf(j)
+                        index(k) = ibuf(j)
+                        j = j - 1
+                        if ( j < 0 ) exit merge_upper
+                    else
+                        array(k) = array(i)
+                        index(k) = index(i)
+                        i = i - 1
+                        if ( i < 0 ) then
+                            array(0:k-1) = buf(0:j)
+                            index(0:k-1) = ibuf(0:j)
+                            exit merge_upper
+                        end if
+                    end if
+                end do merge_upper
+            end if
+        end subroutine merge
+
+
+        pure subroutine reverse_segment( array, index )
+! Reverse a segment of an array in place
+            character(len=*), intent(inout)  :: array(0:)
+            integer(int_size), intent(inout) :: index(0:)
+
+            integer(int_size)         :: itemp, lo, hi
+            character(len=len(array)) :: temp
+
+            lo = 0
+            hi = size( array, kind=int_size ) - 1
+            do while( lo < hi )
+                temp = array(lo)
+                array(lo) = array(hi)
+                array(hi) = temp
+                itemp = index(lo)
+                index(lo) = index(hi)
+                index(hi) = itemp
+                lo = lo + 1
+                hi = hi - 1
+            end do
+
+        end subroutine reverse_segment
+
+    end subroutine char_sort_index
+
+    module subroutine string_sort_index( array, index, work, iwork, reverse )
+! A modification of `string_ord_sort` to return an array of indices that
```

Comment:
Include mention that is also sorts the input `array`

Reply:
Done as noted above.